src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs

```diff
@@ -88,7 +88,12 @@ private static VersionInfo GetVersionInfo()
 
         public override DataViewSchema OutputSchema => _bindings.Schema;
 
-        bool ICanSaveOnnx.CanSaveOnnx(OnnxContext ctx) => _mapper is ICanSaveOnnx onnxMapper ? onnxMapper.CanSaveOnnx(ctx) : false;
+        bool ICanSaveOnnx.CanSaveOnnx(OnnxContext ctx)
+        {
+            const int minimumOpSetVersion = 9;
+            Contracts.Assert(ctx.GetOpSetVersion() >= minimumOpSetVersion, "OpSet version " + ctx.GetOpSetVersion() + " is older than " + LoaderSignature + "'s minimum OpSet version requirement: " + minimumOpSetVersion);
+            return _mapper is ICanSaveOnnx onnxMapper ? onnxMapper.CanSaveOnnx(ctx) : false;
```

Comment:
Another suggestion: The actual exporting to onnx functionality happens in SaveAsOnnx. The operators used are also mentioned in those functions and it would be good keep the version number to be local there. 
E.g. here the RowToRowMapperTransform applies to a lot of transformers and it is not very meaningful to do that check here. 

Can you please move these checks to the individual SaveAsOnnxCore and ISingleCanSaveOnnx functions?

Reply:
I have a different suggestion: Wouldn't it be cleaner to add a method in the `OnnxContext` class called `CheckVersion` or something similar, to do his check to `OnnxContext`? That way the individual components would just have to call it, and pass the minimum version needed, and the name of the component for the error message.
Also, since it seems that most of the components have 9 as the minimum version, I would put a check for version >=9 in the public facing API method that exports to ONNX, or in the `OnnxContextImpl` constructor, and then just the components that require a higher version would need to call `ctx.CheckVersion`.

---
In reply to: [432712627](https://github.com/dotnet/machinelearning/pull/5176#discussion_r432712627) [](ancestors = 432712627)

Reply:
Another suggestion: When you pass the component name for the error message, It may be more informative to pass the specific name of the ONNX operator that requires the higher version instead of the name of the ML.NET component.

---
In reply to: [432946366](https://github.com/dotnet/machinelearning/pull/5176#discussion_r432946366) [](ancestors = 432946366,432712627)

Reply:
I agree with putting a check for versions <= 9 directly on the public API, informing the user we don't support anything below 9. And the idea of the `CheckVersion` method sounds cleaner.
On the other hand, regarding your suggestion of including the operator instead of the ML.NET component on the exception message, I think that from an user perspective knowing which Onnx operator fails isn't very informative as they wouldn't be able to change right away the implementation of the SaveOnnx method that's causing them problems. On the other hand, knowing which component failed to save with the opset they provided might be more useful as they could evaluate if their pipeline truly needs that component, and if they can explore alternatives to it.

---
In reply to: [432946603](https://github.com/dotnet/machinelearning/pull/5176#discussion_r432946603) [](ancestors = 432946603,432946366,432712627)

Reply:
added a checkversion() method

Reply:
Thanks for adding the check for opset>=9 in the `OnnxContextImpl` constructor.
I think all the checks in the different implementations of `ISaveAsOnnx` that check for opset>=9 are redundant now. I would remove them, and leave only the checks that check for a higher version.

---
In reply to: [432972062](https://github.com/dotnet/machinelearning/pull/5176#discussion_r432972062) [](ancestors = 432972062,432946603,432946366,432712627)